src/udev/udev-node.c

```diff
@@ -130,110 +172,394 @@ static int link_find_prioritized(sd_device *dev, bool add, const char *stackdir,
 
         util_path_encode(slink + STRLEN("/dev"), name_enc, sizeof(name_enc));
-        dirname = path_join("/run/udev/links/", name_enc);
+        dirname = path_join(LINKS_DIRNAME, name_enc);
         if (!dirname)
                 return log_oom();
-        filename = path_join(dirname, id_filename);
-        if (!filename)
-                return log_oom();
 
-        if (!add && unlink(filename) == 0)
-                (void) rmdir(dirname);
+        r = device_get_devlink_priority(dev, &priority);
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Failed to get devlink prio: %m");
+
+        if (add) {
+                mkdirat(links_fd, name_enc, 0755);
+                dfd = openat(links_fd, name_enc, O_RDONLY|O_DIRECTORY);
+                if (dfd == -1)
+                        return log_device_error_errno(dev, errno,
+                                                      "Failed to open %s: %m", dirname);
+                create_target_entry(dfd, priority, id_filename, slink);
+        } else {
+                dfd = openat(links_fd, name_enc, O_RDONLY|O_DIRECTORY);
+                if (dfd == -1 && errno != ENOENT)
+                        return log_device_error_errno(dev, errno,
+                                                      "Failed to open %s: %m", dirname);
+                delete_target_entry(dfd, priority, id_filename, slink);
+        }
+
+        dir = fdopendir(dfd);
+        if (!dir) {
+                close(dfd);
+                return log_device_error_errno(dev, errno,
+                                              "Failed to fdopendir %d: %m", dfd);
+        }
 
-        r = link_find_prioritized(dev, add, dirname, &target);
-        if (r < 0) {
-                log_device_debug(dev, "No reference left, removing '%s'", slink);
+        semidx = get_sema_index(slink);
+        lock_failed = lock_slink(semid, semidx);
+        if (lock_failed)
+                log_error_errno(errno, "Failed to lock %s: %m", slink);
+
+        r = link_find_prioritized(dev, add, dir, slink, &target);
+        if (r == TARGET_NEEDS_CLEANUP) {
+                cleanup_old_targets(dirname, dev);
+                r = link_find_prioritized(dev, add, dir, slink, &target);
+                /* A single cleanup must be enough */
+                assert(r != TARGET_NEEDS_CLEANUP);
+        }
+        if (r != TARGET_FOUND) {
+                log_debug("No reference left, remove '%s'", slink);
                 if (unlink(slink) == 0)
                         (void) rmdir_parents(slink, "/");
-        } else
-                (void) node_symlink(dev, target, slink);
+        } else {
+                log_debug("Creating link '%s' to '%s'", slink, target);
+                mkdir_parents(slink, 0755);
+                node_symlink(dev, target, slink);
+        }
 
-        if (add)
-                do {
-                        _cleanup_close_ int fd = -1;
+        if (!lock_failed)
```

Comment:
`if (lock_failed == 0)`. Yeah, I prefer `if (ret_lock == 0)`...

Reply:
I've a different taste here. I hope this is not a blocker criterion for you.